File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.*;
import java.util.*;

public class Main{

    static BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw=new BufferedWriter(new OutputStreamWriter(System.out));

    static int N;
    static int arr[][];
    static int cnt=0;
    static boolean visited[][];
    static ArrayList<Integer> bag=new ArrayList<>();

    static int []dx={1,-1,0,0};
    static int []dy={0,0,-1,1};



    public static void main(String[] args) throws IOException {

        N=Integer.parseInt(br.readLine());
        arr=new int[N][N];
        visited = new boolean[N][N];

        for(int i=0;i<N;i++){

            String str=br.readLine();

            for(int j=0;j<N;j++){

                arr[i][j]=str.charAt(j)-'0';



            }
        }

        for(int i=0;i<N;i++){
            for(int j=0;j<N;j++){
                if(visited[i][j]!=true && arr[i][j]==1)BFS(i,j);
            }
        }
        bw.write(String.valueOf(cnt)+"\n");

        Collections.sort(bag);

        for(int i=0 ; i<bag.size();i++){
            bw.write(String.valueOf(bag.get(i))+"\n");
        }
        
        bw.flush();
        bw.close();










    }

    static void BFS(int x,int y){

        cnt++;
        Queue<Node>queue=new LinkedList<>();
        queue.add(new Node(x,y));
        visited[x][y]=true;
        int  count=1;
        while(!queue.isEmpty()){

            Node cur=queue.poll();
            for(int i=0;i<4;i++){

                int nx=cur.x+dx[i];
                int ny=cur.y+dy[i];
                if(check(nx,ny))continue;   //배열범위를 넘어갔을경우
                if(arr[nx][ny]!=1)continue;  //1이 아닐경우
                if(visited[nx][ny])continue; // 이미방문했을경우

                queue.add(new Node(nx,ny));
                visited[nx][ny]=true;
                count++;


            }


        }
        bag.add(count);




    }

    static boolean check(int x,int y){

        if(x<0 || x>=N || y<0 || y>=N)return true;
        return false;
    }

    static class Node{
        int x;
        int y;

        public Node(int x, int y) {
            this.x = x;
            this.y = y;
        }
    }


}

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
import sys
from collections import deque
movex=[-1,0,1,0]
movey=[0,1,0,-1]
cnt=0
count=0

bag=[]

N=int(input())
graph=[[] for _ in range(N)]

for i in range(N):
    graph[i]=list(map(int,input().strip()))


visited=[[0] * N for i in range (N)]

def BFS(x,y) :
    count=0
    visited[x][y]=1
    queue=deque()
    queue.append([x,y])
    count+=1
    while(queue) :
        cur=queue.popleft()
        for i in range(4):
            nx = cur[0] + movex[i]
            ny = cur[1] + movey[i]
            if (nx < 0 or nx >= N or ny < 0 or ny >= N): continue
            if (visited[nx][ny] != 0 or graph[nx][ny] == 0): continue
            queue.append([nx,ny])
            visited[nx][ny]=1
            count+=1
    bag.append(count)


for x in range(N):

    for y in range(N) :
        if(graph[x][y]==1 and visited[x][y]==0) :
            BFS(x,y)
            cnt+=1

bag.sort()


print(cnt)
for i in bag :
    print(i)
